pkg/sentry/fsimpl/fuse/read_write.go

```diff
@@ -0,0 +1,161 @@
+		maxPages = uint32(fs.conn.maxPages)
+	}
+
+	t := kernel.TaskFromContext(ctx)
+	if t == nil {
+		log.Warningf("fusefs.Read: couldn't get kernel task from context")
+		return nil, 0, syserror.EINVAL
+	}
+
+	var outs [][]byte
+	var sizeRead uint32
+
+	// readSize is a multiple of usermem.PageSize.
+	// Always request bytes as a multiple of pages.
+	pagesRead, pagesToRead := uint32(0), uint32(readSize/usermem.PageSize)
+
+	// This loop is intended for fragmented read where the bytes to read is
+	// larger than either the maxPages or maxRead.
+	// For the majority of reads with normal size, this loop should only
+	// execute once.
+	for pagesRead < pagesToRead {
+		pagesCanRead := pagesToRead - pagesRead
+		if pagesCanRead > maxPages {
+			pagesCanRead = maxPages
+		}
+
+		in := linux.FUSEReadIn{
+			Fh:        fd.Fh,
+			Offset:    off + (uint64(pagesRead) << usermem.PageShift),
+			Size:      pagesCanRead << usermem.PageShift,
+			LockOwner: 0, // TODO(gvisor.dev/issue/3245): file lock
+			ReadFlags: 0, // TODO(gvisor.dev/issue/3245): |= linux.FUSE_READ_LOCKOWNER
+			Flags:     fd.statusFlags(),
+		}
+
+		req, err := fs.conn.NewRequest(auth.CredentialsFromContext(ctx), uint32(kernel.TaskFromContext(ctx).ThreadID()), fd.inode().NodeID, linux.FUSE_READ, &in)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		// TODO(gvisor.dev/issue/3247): support async read.
+
+		res, err := fs.conn.Call(t, req)
+		if err != nil {
+			return nil, 0, err
+		}
+		if err := res.Error(); err != nil {
+			return nil, 0, err
+		}
+
+		// Not enough bytes in response,
+		// either we reached EOF,
+		// or the FUSE server sends back a response
+		// that cannot even fit the hdr.
+		if len(res.data) <= res.hdr.SizeBytes() {
+			if len(res.data) < res.hdr.SizeBytes() {
+				return nil, 0, fmt.Errorf("FUSE_READ response too small. Minimum length required: %d,  but got length %d", res.hdr.SizeBytes(), len(res.data))
+			}
+			break
+		}
+
+		// Directly using the slice to avoid extra copy.
+		out := res.data[res.hdr.SizeBytes():]
+		outs = append(outs, out)
+		sizeRead += uint32(len(out))
+
+		pagesRead += pagesCanRead
+	}
+
+	defer fs.ReadCallback(ctx, fd, off, size, sizeRead, attributeVersion)
+
+	// No bytes returned: offset >= EOF
+	if len(outs) == 0 {
+		return []byte{}, 0, io.EOF
+	}
+
+	// Finished with one reply.
+	// Return the slice directly from the buffer in response.
+	if len(outs) == 1 {
+		return outs[0], sizeRead, nil
+	}
+
+	// Join data from multiple fragmented replies.
+	// TODO(gvisor.dev/issue/3628): avoid this extra copy,
+	// perhaps we can use iovec, which requires upperstream support.
+	buf := make([]byte, sizeRead)
+	bufPos := 0
+	for _, v := range outs {
+		bufPos += copy(buf[bufPos:], v)
+	}
+
+	return buf, sizeRead, nil
+}
+
+// ReadCallback updates several information after receiving a read response.
+func (fs *filesystem) ReadCallback(ctx context.Context, fd *regularFileFD, off uint64, size uint32, sizeRead uint32, attributeVersion uint64) {
```

Comment:
[nit]: Naming is unclear - what does this functon do? Update Inode attributes right of the read file? Maybe rename to `UpdateAttr`?

Reply:
See all the todo. It should be processing the reply of a FUSE_READ request and update all the odds and ends